### ISSUE TITLE
GraphQLServlet doesn't play nice with custom DataFetchers

### ISSUE THREAD

**Issue by tomshackell** [graphql-java/graphql-java-servlet#28]
The GraphQLServlet does not allow proper use of the 'context' item that is needed for custom DataFetchers. This (executable) example illustrates the problem:

```
import java.util.Arrays;
import java.util.List;

import graphql.ExecutionResult;
import graphql.GraphQL;
import graphql.Scalars;
import graphql.schema.GraphQLList;
import graphql.schema.GraphQLObjectType;
import graphql.schema.GraphQLSchema;

public class GraphQLTest {
    public static void main(String[] args) {
        try {
            // build schema
            GraphQLObjectType personType = GraphQLObjectType.newObject()
                    .name("Person")
                    .field(f -> f.name("name")
                                 .type(Scalars.GraphQLString)
                                 .dataFetcher(e -> ((Person) e.getSource()).name))
                    .field(f -> f.name("age")
                                 .type(Scalars.GraphQLInt)
                                 .dataFetcher(e -> ((Person) e.getSource()).age))
                    .build();

            GraphQLObjectType teamType = GraphQLObjectType.newObject()
                    .name("Team")
                    .field(f -> f.name("people")
                                 .type(new GraphQLList(personType))
                                 .dataFetcher(e -> ((Team) e.getSource()).people))
                    .build();

            GraphQLSchema schema = GraphQLSchema.newSchema()
                    .query(teamType).build();

            // create our 'database' that implements the schema
            Team team = new Team(Arrays.asList(
                    new Person("Alice", 24), new Person("Bob", 45), 
                    new Person("Carol", 33), new Person("David", 28)));

            // run the query
            ExecutionResult result = new GraphQL(schema)
                    .execute("{ people { name, age } }", team);
            System.out.println(result.getData().toString());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static class Person {
        public final String name;
        public final int age;

        public Person(String name, int age) { this.name = name; this.age = age; }
    }

    private static class Team {
        public final List<Person> people;

        public Team(List<Person> people) { this.people = people; }
    }
}
```

This is an idiomatic way to use data fetchers.

The problem is that the graphql-java-servlet implementation provides no way to provide the 'context' item to the execute method ('team' in this case). Indeed looking at the source code it seems that the 'context' parameter to execute is being used to pass something else (a GraphQLContext). This will not work with custom DataFetchers that make use of getSource (as most do). 

In the above example, the GraphQLServlet implementation effectively does this:

`ExecutionResult result = new GraphQL(schema).execute("{ people { name, age } }", 
   new GraphQLContext(...));`

Which will crash:

`java.lang.ClassCastException: graphql.servlet.GraphQLContext cannot be cast to GraphQLTest$Team`

on this line:

`.dataFetcher(e -> ((Team) e.getSource()).people))`

**Comment by apottere** [comment]
I think this is the intended behavior, from the javadocs on `.getSource()`:
```
@return the current object being queried
```

In practice, this value can't be relied on for root-level queries, since there isn't an object being queried.  Once the root object has been fetched (in this case, a list of `Person`), then each field on nested objects will get the java object that's being queried as their source.  In this example, you're actually registering `Team` as your root Query type, so it doesn't make sense to use `.getSource()`.  It looks like graphql-java passes the context as the source for root-level queries, which personally I think just causes more confusion, since it's already available by `.getContext()`.

That being said, if you need data sources globally available to your query (e.g. for resolving root objects), then you should put them on the context and pull them off the context.  The servlet does allow you to use your own context object.  If you're using OSGI there's an interface you can implement, and if you're not you can subclass the servlet and override `createContext`.

**Comment by tomshackell** [comment]
It would definitely be good to check with the GraphQL people what the intent was, as I agree it's not documented at all.

What is certainly true is that other parts of the graphql-java ecosystem do rely on the source of the root-level query being the context object.

I came across this issue using graphql-java-annotations, which constructs the schema automatically from annotations on a Java type. The schema that graphql-java-annotation's builds does assume that you're going to pass an instance of the class you've used to build the annotation as the context, and it will blow up if you don't. This means that currently graphql-java-servlet is incompatible with graphql-java-annotations.

**Comment by apottere** [comment]
Interesting, I'll follow up on this.

**Comment by apottere** [comment]
This is the relevant PR on graphql-java: https://github.com/graphql-java/graphql-java/issues/211

**Comment by apottere** [comment]
This is now fixed as of e9a6204.  You can pass a `StaticGraphQLRootObjectBuilder` that has the "root" object for graphql-java-annotations to the servlet and it should work.
